File: output_20260318_123118/gpt52/c/Stack/filter.c
#include "stdlib.h"

struct node
{
    struct node *next;
    int value;
};

struct stack
{
    struct node *head;
};

/*@

predicate nodes(struct node *n) =
    n == 0 ?
        emp
    :
        n->next |-> ?next &*& n->value |-> ?v &*& malloc_block_node(n) &*& nodes(next);

predicate stack(struct stack *s) =
    s->head |-> ?h &*& malloc_block_stack(s) &*& nodes(h);

@*/

struct stack *create_stack()
    //@ requires true;
    //@ ensures stack(result);
{
    struct stack *stack = malloc(sizeof(struct stack));
    if (stack == 0)
    {
        abort();
    }
    stack->head = 0;
    
    
    return stack;
}

void stack_push(struct stack *stack, int value)
    //@ requires stack(stack);
    //@ ensures stack(stack);
{
    
    struct node *n = malloc(sizeof(struct node));
    if (n == 0)
    {
        abort();
    }
    n->next = stack->head;
    n->value = value;
    stack->head = n;
    
    
}

int stack_pop(struct stack *stack)
    //@ requires stack(stack) &*& stack->head |-> ?h &*& h != 0;
    //@ ensures stack(stack);
{
    
    struct node *head = stack->head;
    
    int result = head->value;
    stack->head = head->next;
    free(head);
    
    return result;
}

typedef bool int_predicate(int x);

/*@

predicate int_predicate_pred(int_predicate *p) = is_int_predicate(p);

@*/

struct node *nodes_filter(struct node *n, int_predicate *p)
    //@ requires nodes(n) &*& int_predicate_pred(p);
    //@ ensures nodes(result) &*& int_predicate_pred(p);
{
    if (n == 0)
    {
        return 0;
    }
    else
    {
        
        bool keep = p(n->value);
        if (keep)
        {
            struct node *next = nodes_filter(n->next, p);
            
            
            n->next = next;
            
            return n;
        }
        else
        {
            struct node *next = n->next;
            free(n);
            struct node *result = nodes_filter(next, p);
            return result;
        }
    }
}

void stack_filter(struct stack *stack, int_predicate *p)
    //@ requires stack(stack) &*& int_predicate_pred(p);
    //@ ensures stack(stack) &*& int_predicate_pred(p);
{
    
    struct node *head = nodes_filter(stack->head, p);
    
    stack->head = head;
    
    
    
}

void nodes_dispose(struct node *n)
    //@ requires nodes(n);
    //@ ensures emp;
{
    
    if (n != 0)
    {
        nodes_dispose(n->next);
        free(n);
    }
}

void stack_dispose(struct stack *stack)
    //@ requires stack(stack);
    //@ ensures emp;
{
    
    nodes_dispose(stack->head);
    free(stack);
}

bool neq_20(int x) 
    //@ requires true;
    //@ ensures true;
{
    return x != 20;
}

int main()
    //@ requires true;
    //@ ensures true;
{
    struct stack *s = create_stack();
    stack_push(s, 10);
    stack_push(s, 20);
    stack_push(s, 30);
    //@ close int_predicate_pred(neq_20);
    stack_filter(s, neq_20);
    //@ open int_predicate_pred(neq_20);
    stack_dispose(s);
    return 0;
}

File: output_log/claude-opus/sample_5/rust/qwen3/rust/purely_unsafe/solutions/filter.rs
use std::alloc::{alloc, dealloc, handle_alloc_error, Layout};

#[repr(C)]
struct Node {
    value: i32,
    next: *mut Node,
}

#[repr(C)]
struct Stack {
    head: *mut Node,
}

type I32Predicate = fn(i32) -> bool;

#[predicate]
#[verifier(public)]
fn node_pred(n: *mut Node) = n != 0 &*& malloc_block_Node(n) &*& (*n).value |-> _ &*& (*n).next |-> _ &*&
    (let next = (*n).next; let value = (*n).value; node_pred(next) &*& value == value) || // recursive
    n == 0 &*& emp;

#[predicate]
#[verifier(public)]
fn list_pred(head: *mut Node) =
    head == 0 ? emp : node_pred(head) &*& list_pred((*head).next);

#[predicate]
#[verifier(public)]
fn stack_pred(s: *mut Stack) = s != 0 &*& malloc_block_Stack(s) &*& (*s).head |-> ?h &*& list_pred(h);

#[lemma]
#[verifier(public)]
fn node_pred_split(n: *mut Node)
    requires node_pred(n),
    ensures (*n).next |-> ?next &*& (*n).value |-> ?value &*& node_pred(next)
{
    open node_pred(n);
    assert (*n).next |-> next;
    assert (*n).value |-> value;
    close node_pred(n);
}

#[lemma]
#[verifier(public)]
fn list_pred_split(head: *mut Node)
    requires list_pred(head),
    ensures head == 0 ? emp : node_pred(head) &*& list_pred((*head).next)
{
    open list_pred(head);
    close list_pred(head);
}

unsafe fn filter_nodes(n: *mut Node, p: I32Predicate) -> *mut Node
#[requires list_pred(n)]
#[ensures list_pred(result)]
{
    if n.is_null() {
        std::ptr::null_mut()
    } else {
        open list_pred(n);
        open node_pred(n);
        let keep = p((*n).value);
        let next;
        let res;
        if keep {
            next = filter_nodes((*n).next, p);
            (*n).next = next;

            close node_pred(n);
            close list_pred(n);
            n
        } else {
            next = (*n).next;
            dealloc(n as *mut u8, Layout::new::<Node>());
            res = filter_nodes(next, p);
            close list_pred(res);
            res
        }
    }
}

unsafe fn dispose_nodes(n: *mut Node)
#[requires list_pred(n)]
#[ensures emp]
{
    if !n.is_null() {
        open list_pred(n);
        open node_pred(n);
        dispose_nodes((*n).next);
        dealloc(n as *mut u8, Layout::new::<Node>());
    }
}

impl Stack {
    unsafe fn create() -> *mut Stack
    #[requires emp]
    #[ensures stack_pred(result)]
    {
        let stack = alloc(Layout::new::<Stack>()) as *mut Stack;
        if stack.is_null() {
            handle_alloc_error(Layout::new::<Stack>());
        }
        (*stack).head = std::ptr::null_mut();

        close list_pred(std::ptr::null_mut());
        close stack_pred(stack);

        stack
    }

    unsafe fn push(stack: *mut Stack, value: i32)
    #[requires stack_pred(stack)]
    #[ensures stack_pred(stack)]
    {
        open stack_pred(stack);
        open list_pred((*stack).head);

        let n = alloc(Layout::new::<Node>()) as *mut Node;
        if n.is_null() {
            handle_alloc_error(Layout::new::<Node>());
        }
        (*n).next = (*stack).head;
        (*n).value = value;

        close node_pred(n);
        close list_pred(n);
        (*stack).head = n;

        close stack_pred(stack);
    }

    unsafe fn pop(stack: *mut Stack) -> i32
    #[requires stack_pred(stack) &*& (*(*stack).head).value |-> ?v]
    #[ensures stack_pred(stack)]
    {
        open stack_pred(stack);
        open list_pred((*stack).head);
        open node_pred((*stack).head);
        assert (*stack).head != std::ptr::null_mut();

        let head = (*stack).head;

        let result = (*head).value;
        (*stack).head = (*head).next;
        dealloc(head as *mut u8, Layout::new::<Node>());

        close list_pred((*stack).head);
        close stack_pred(stack);

        result
    }

    unsafe fn filter(stack: *mut Stack, p: I32Predicate)
    #[requires stack_pred(stack)]
    #[ensures stack_pred(stack)]
    {
        open stack_pred(stack);
        open list_pred((*stack).head);

        let head = filter_nodes((*stack).head, p);

        (*stack).head = head;

        close stack_pred(stack);
    }

    unsafe fn dispose(stack: *mut Stack)
    #[requires stack_pred(stack)]
    #[ensures emp]
    {
        open stack_pred(stack);
        dispose_nodes((*stack).head);
        dealloc(stack as *mut u8, Layout::new::<Stack>());
    }
}

unsafe fn neq_20(x: i32) -> bool {
    x != 20
}

fn main() {
    unsafe {
        let s = Stack::create();
        Stack::push(s, 10);
        Stack::push(s, 20);

        Stack::filter(s, neq_20);
        Stack::dispose(s);
    }
}